src/eclic/mode.rs

```diff
@@ -0,0 +1,364 @@
+            ".align 6",
+            ".option pop",
+            "addi sp, sp, -16*REGBYTES",
+            //
+            "sw ra, 0*REGBYTES(sp)",
+            "sw t0, 1*REGBYTES(sp)",
+            "sw t1, 2*REGBYTES(sp)",
+            "sw t2, 3*REGBYTES(sp)",
+            "sw t3, 4*REGBYTES(sp)",
+            "sw t4, 5*REGBYTES(sp)",
+            "sw t5, 6*REGBYTES(sp)",
+            "sw t6, 7*REGBYTES(sp)",
+            "sw a0, 8*REGBYTES(sp)",
+            "sw a1, 9*REGBYTES(sp)",
+            "sw a2, 10*REGBYTES(sp)",
+            "sw a3, 11*REGBYTES(sp)",
+            "sw a4, 12*REGBYTES(sp)",
+            "sw a5, 13*REGBYTES(sp)",
+            "sw a6, 14*REGBYTES(sp)",
+            "sw a7, 15*REGBYTES(sp)",
+            //
+            "add a0, sp, zero",
+            "jal ra, _start_trap_rust",
+            //
+            "lw ra, 0*REGBYTES(sp)",
+            "lw t0, 1*REGBYTES(sp)",
+            "lw t1, 2*REGBYTES(sp)",
+            "lw t2, 3*REGBYTES(sp)",
+            "lw t3, 4*REGBYTES(sp)",
+            "lw t4, 5*REGBYTES(sp)",
+            "lw t5, 6*REGBYTES(sp)",
+            "lw t6, 7*REGBYTES(sp)",
+            "lw a0, 8*REGBYTES(sp)",
+            "lw a1, 9*REGBYTES(sp)",
+            "lw a2, 10*REGBYTES(sp)",
+            "lw a3, 11*REGBYTES(sp)",
+            "lw a4, 12*REGBYTES(sp)",
+            "lw a5, 13*REGBYTES(sp)",
+            "lw a6, 14*REGBYTES(sp)",
+            "lw a7, 15*REGBYTES(sp)",
+            //
+            "addi sp, sp, 16*REGBYTES",
+            "mret",
+        );
+    }
+}
+
+// FIXME: defining this with global_asm causes linked code that defines
+// an interrupt vector to raise a "<VECTOR> changed binding to STB_GLOBAL"
+//
+// As-is, defining a vector raises a "<VECTOR> changed binding to STB_WEAK"
+//
+// The vector also only seems to be called once...
+#[link_section = ".text"]
+#[no_mangle]
+fn vector_symbols() {
+    unsafe {
+        asm!(
+            ".weak INT_SFT",
+            ".weak INT_TMR",
+            ".weak INT_BWEI",
+            ".weak INT_PMOVI",
+            ".weak WWDGT",
+            ".weak EXTI_LVD",
+            ".weak TAMPER",
+            ".weak RTC",
+            ".weak FMC",
+            ".weak RCU",
+            ".weak EXTI_LINE0",
+            ".weak EXTI_LINE1",
+            ".weak EXTI_LINE2",
+            ".weak EXTI_LINE3",
+            ".weak EXTI_LINE4",
+            ".weak DMA0_CHANNEL0",
+            ".weak DMA0_CHANNEL1",
+            ".weak DMA0_CHANNEL2",
+            ".weak DMA0_CHANNEL3",
+            ".weak DMA0_CHANNEL4",
+            ".weak DMA0_CHANNEL5",
+            ".weak DMA0_CHANNEL6",
+            ".weak ADC0_1",
+            ".weak CAN0_TX",
+            ".weak CAN0_RX0",
+            ".weak CAN0_RX1",
+            ".weak CAN0_EWMC",
+            ".weak EXTI_LINE9_5",
+            ".weak TIMER0_BRK",
+            ".weak TIMER0_UP",
+            ".weak TIMER0_TRG_CMT",
+            ".weak TIMER0_CHANNEL",
+            ".weak TIMER1",
```

Comment:
Raises a warning when compiling [longan-nano/examples/interrupt](https://github.com/riscv-rust/longan-nano/blob/master/examples/interrupt.rs)

```
warning: TIMER1 changed binding to STB_WEAK
   |
note: instantiated into assembly here
  --> <inline asm>:33:1
   |
33 | .weak TIMER1
```

The timer interrupt fires, but only once at the beginning of execution.

Not sure how to solve this...

Using `global_asm` raises a `TIMER1 changed binding to STB_GLOBAL`  error, and the build fails.

Reply:
It should be `global_asm` because declaring it as a function may add a few instructions in the beginning and all the handler addresses will become shifted in this case. In any case, this is not code, so should not be callable.

Reply:
> This part and vectors should be global_asm because declaring it as a function may add a few instructions in the beginning and all the handler addresses will become shifted in this case.

I completely agree, but when I tried to make it `global_asm` it raises an error in the `longan-nano` example code:

```
error: TIMER1 changed binding to STB_GLOBAL
```

That error occurs both with the `interrupt` macro from `gd32vf103-pac`, and the current code from the `master` branch.

Do you think it would help to implement a hand-written proc-macro, similar to what's in `cortex-m` and `avr-device`?

Reply:
Something is strange here. I got different results for debug and release builds:
```
% nm target/riscv32imac-unknown-none-elf/debug/examples/interrupt | grep " TIMER1"
08000568 T TIMER1
% nm target/riscv32imac-unknown-none-elf/release/examples/interrupt | grep " TIMER1"
080001c2 W TIMER1
```
It's ok when STB_GLOBAL is used (like in debug here), but STB_WEAK looks wrong. Maybe each interrupt handler needs a default implementation like was done in `stm32f1`, for example. It provides `device.x` with a lot of lines of the following format `PROVIDE(TIM4 = DefaultHandler);` and then an array is created with all the handlers: https://docs.rs/stm32f1/latest/src/stm32f1/stm32f103/mod.rs.html#82-177

Reply:
Moved everything into a `global_asm` block, and use a `device.x` file in the `longan-nano` example.

Using the `device.x` file to define the default handlers clears the `STB_WEAK/GLOBAL` errors, but the blinky still stays static (constantly on after 1 second). Not sure what is causing the error..